File: euler_064.py
from time import time
from math import sqrt, floor

def len_cf_of_square_root(n):
    if sqrt(n) == floor(sqrt(n)):
        return 0
    a = [floor(sqrt(n))]
    b = [0, a[0]]
    c = [0, n - a[0]**2]
    i = 1
    while a[-1] != 2 * a[0]:
        i += 1
        a.append(floor((a[0] + b[i-1]) / c[i-1]))
        b.append(a[i-1] * c[i-1] - b[i-1])
        c.append((n - b[i]**2) / c[i-1])
    return len(a) - 1

start = time()

odd = 0

for n in range(1,10001):
    p = len_cf_of_square_root(n)
    if p%2 == 1:
        odd += 1

print("Number of odd-period CFs: " + str(odd))


stop = time()
print("Time: " + str(stop-start))

File: euler_064.js
import { continuedFractionGenerator, simpleSqrtCF } from "./continuedFractions.mjs";
import Decimal from "./decimal.mjs";

Decimal.set({precision: 100});

const start = Date.now();

const period = (n) => {
    if (new Decimal(n).sqrt().isInt()) {
        return 0;
    }
    let a = [new Decimal(n).sqrt().floor()];
    let b = [new Decimal(0), a[0]];
    let c = [b[0], new Decimal(n).minus(a[0].toPower(2))];

    let i = 1;
    while (!(new Decimal(a[0].times(2)).equals(a[a.length - 1]))) {
        i++;
        a.push(((a[0].plus(b[i - 1])).dividedBy(c[i - 1])).floor());
        b.push((a[i - 1].times(c[i - 1])).minus(b[i - 1]));
        c.push((new Decimal(n).minus(b[i].toPower(2))).dividedBy(c[i - 1]));
    }
    return a.length - 1;
}

let oddPeriods = 0;
for (let k = 2; k < 10001; k++) {
    if (period(k) % 2 === 1) {
        oddPeriods++;
    }
}

const end = Date.now();

console.log("Execute time: ", end - start, " Number of odd-period continued fractions is: ", oddPeriods);